File: app/src/main/java/fup/prototype/robprototype/view/activities/BaseActivity.java
package fup.prototype.robprototype.view.activities;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AppCompatActivity;
import fup.prototype.robprototype.R;
import fup.prototype.robprototype.view.fragments.BaseFragment;

public abstract class BaseActivity extends AppCompatActivity {

    private static final String KEY_FRAGMENT = "keyFragment";

    private BaseFragment baseFragment;

    @Override
    protected void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
        baseFragment = getOrCreateFragment(savedInstanceState);
        initContent();
    }

    private BaseFragment getOrCreateFragment(@Nullable final Bundle savedInstanceState) {
        if (savedInstanceState != null) {
            final String keyFragment = savedInstanceState.getString(KEY_FRAGMENT);
            return (BaseFragment) getSupportFragmentManager().getFragment(savedInstanceState, keyFragment);
        } else {
            return createInitialContentFragment();
        }
    }

    private void initContent() {
        final FragmentManager fragmentManager = getSupportFragmentManager();
        final FragmentTransaction transaction = fragmentManager.beginTransaction();
        transaction.replace(R.id.fragment_container, baseFragment);
        transaction.commit();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(KEY_FRAGMENT, baseFragment.getKey());
        getSupportFragmentManager().putFragment(outState, baseFragment.getKey(), baseFragment);
    }

    protected abstract int getLayoutId();

    protected abstract BaseFragment createInitialContentFragment();
}

File: app/src/main/java/com/rxdroid/app/view/base/BaseActivity.kt
package com.rxdroid.app.view.base

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import com.rxdroid.app.R
import com.rxdroid.app.view.base.fragments.BaseFragment

abstract class BaseActivity : AppCompatActivity() {

    private object Constants {
        const val KEY_FRAGMENT = "keyFragment"
    }

    private var baseFragment: BaseFragment<*>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutId())
        baseFragment = getOrCreateFragment(savedInstanceState)
        initContent()
    }

    private fun getOrCreateFragment(savedInstanceState: Bundle?): BaseFragment<*> {
        savedInstanceState?.let {
            val keyFragment = savedInstanceState.getString(Constants.KEY_FRAGMENT)!!
            return supportFragmentManager.getFragment(savedInstanceState, keyFragment) as BaseFragment<*>
        }
        return createInitialContentFragment()
    }

    private fun initContent() {
        val fragmentManager: FragmentManager = supportFragmentManager
        fragmentManager.beginTransaction().replace(R.id.fragment_container, baseFragment as Fragment).commit()
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        super.onSaveInstanceState(outState)
        outState?.let {
            val keyFragment = baseFragment?.getKey()!!
            outState.putString(Constants.KEY_FRAGMENT, keyFragment)
            supportFragmentManager.putFragment(outState, keyFragment, baseFragment as Fragment)
        }
    }

    protected abstract fun createInitialContentFragment(): BaseFragment<*>

    protected abstract fun getLayoutId(): Int
}